server/subscribe.go

```diff
@@ -104,7 +104,24 @@ func (p *Plugin) getChannelsSubscribed(jwh *JiraWebhook) ([]string, error) {
 		return nil, err
 	}
 
-	subIds := subs.Channel.IdByEvent[jwh.WebhookEvent]
+	eventEnums := jwh.toEventEnums()
+
+	// Gather sub ids in a map for performance
+	subIdMap := map[string]int{}
+	for _, enum := range eventEnums {
+		ids := subs.Channel.IdByEvent[enum]
+		for _, id := range ids {
+			subIdMap[id] = 0
+		}
+	}
+
+	// Convert map to slice of sub ids
+	subIds := make([]string, len(subIdMap))
+	i := 0
+	for k := range subIdMap {
+		subIds[i] = k
+		i++
+	}
```

Comment:
Maybe a more general design change, but can we create an eventType -> []subIds map when channels subscribe, rather than rebuild it for every webhook event?

Reply:
There is an `eventType -> []subIds` map, however the `eventType` used is now the more granular string-based enums. Each jira webhook payload will most often be associated with more than one enum. I.e. if an issue's description is updated, we should search for subscriptions that are subscribed to specifically description changes, and search for subscriptions that are subscribed to all updates to the any of the issue's fields. This makes it so we need to find subscriptions that have _either_ of these two subscription enums. So we collect all subscriptions by querying that map with both enums.

This is something that should be explicitly discussed with the team, as it is a significant change in how the subscriptions work. This was one of the proposed solutions @levb spoke of in our talk earlier this week. The performance and code cleanliness implications of my implementation are definitely something to discuss.

Thanks for your review @cpoile!

Reply:
@cpoile Here is my proposed solution, to avoid the upfront excessive event lookup during the webhook call. The events in the `eventType -> subIds[]` should remain to be the high-level `Issue Updated` and `Issue Created` event types, and not the specific fields that the user has chosen. I believe this is what you are getting at with your comment. This works because the specific fields the user has chosen do not matter until we are in the loop checking each subscription in more detail.

So as a user if I choose "Issue Updated - Description" as one of my options and save the subscription, then the backend will ensure that this sub id exists in the map, for the value pointed to by the "Issue Updated" event type. If I uncheck the "Issue Updated - Description" option, and there is not another "Issue Updated - \<field\>" selected in my subscription, then the sub id will be removed from the map.

This also points out to me that tests need to be written for this function in `subscribe.go`, so I'll include them in the PR as well.